tests/pthread_rwlock/main.c

```diff
@@ -55,11 +55,22 @@ static volatile unsigned counter;
 
 static kernel_pid_t main_thread_pid;
 
+/* The test assumes that 'printf/puts' are non interruptible operations
+ * use a mutex to guarantee it */
+static mutex_t stdout_mutex = MUTEX_INIT;
```

Comment:
Couldn't this be generalized? It would imagine more use cases where interrupted prints can screw up a tests. #11818 deactivated ISR to achieve a similar result.

https://github.com/RIOT-OS/RIOT/blob/0440c646913fb37598d70572429ce7aa73ef80b0/examples/suit_update/main.c#L43-L47

Reply:
Deactivated ISR do not work if the test is using DMA printfs.
As we cannot know in RIOT if interrupt are disabled, the dma implementation of `uart_write` cannot do an disabled interrupt safe implementation.

And printing is really slow so disabling interrupts for a long time seems bad.

I was talking about having a `console_lock` in the `dma` print, that could also be useful when you want a block to be printed as a transaction that could even be done in multiple `printf` by explicitly taking the lock.
https://github.com/RIOT-OS/RIOT/issues/10800#issuecomment-455195344
It would need to be a `rlock` though.

Reply:
> I was talking about having a console_lock in the dma print, that could also be useful when you want a block to be printed as a transaction that could even be done in multiple printf by explicitly taking the lock.
> #10800 (comment)
> It would need to be a rlock though.

+1, even if it would need rlock, I think it would make sense to have a feature like this for testing. Do you have plans on working on this soon? If not we can go with the local implementation for now, but I wouldn't want it to start being replicated every time we need to make sure prints are correctly going threw.

Reply:
Edit: in the dma print *discussion*

No plans to work on it soon, I also have no idea how it can be done at the C library level "automatically" :/

Reply:
An implementation could be to define a `console_lock` rmutex and `locked_printf`, `locked_puts` macros or functions (and others).

But it would only solve the case where they are explicitly used, so more in the tests applications. It would be enough as long as there is no other debug outputs. So not when using the `gnrc` debug module at the same time, or drivers putting debug messages for example.

If this is enough I could try putting some time on it.


Managing to make it the default could be another task and split in a dedicated module like "printf_auto_rlock".

Reply:
> An implementation could be to define a console_lock rmutex and locked_printf, locked_puts macros or functions (and others).

> But it would only solve the case where they are explicitly used, so more in the tests applications. It would be enough as long as there is no other debug outputs. So not when using the gnrc debug module at the same time, or drivers putting debug messages for example.

Hmm, thats true. Many of the use cases I can think fall in the `printf_auto_rlock` case. I think it makes more sense working on "printf_auto_rlock", I'll think it over a little more and see if I can find more use cases, but if not we can go with this as is.